File: examples/hitl/go/require_approval_policy.go
// Package main demonstrates creating HITL policies with require_approval action
// and VALIDATES that enforcement actually works via ProxyLLMCall.
//
// This example shows how to create a policy that triggers
// Human-in-the-Loop (HITL) approval using the `require_approval` action.
//
// The `require_approval` action:
// - Enterprise: Pauses execution and creates an approval request in the HITL queue
// - Community: Auto-approves immediately (upgrade path to Enterprise)
//
// Use cases:
// - High-value transaction oversight (EU AI Act Article 14, SEBI AI/ML)
// - Admin access detection
// - Sensitive data access control
//
// VALIDATION: This example exits with code 1 if any assertion fails.
package main

import (
	"fmt"
	"os"
	"strings"
	"time"

	axonflow "github.com/getaxonflow/axonflow-sdk-go/v5"
)

var failures []string

func assertCheck(condition bool, message string) {
	if condition {
		fmt.Printf("   ✓ PASS: %s\n", message)
	} else {
		fmt.Printf("   ❌ FAIL: %s\n", message)
		failures = append(failures, message)
	}
}

func main() {
	// Initialize the client (ClientID is used as tenant ID for policy APIs)
	agentURL := os.Getenv("AXONFLOW_ENDPOINT")
	if agentURL == "" {
		agentURL = os.Getenv("AXONFLOW_AGENT_URL")
	}
	if agentURL == "" {
		agentURL = "http://localhost:8080"
	}
	clientID := os.Getenv("AXONFLOW_CLIENT_ID")
	if clientID == "" {
		clientID = "demo-tenant"
	}
	clientSecret := os.Getenv("AXONFLOW_CLIENT_SECRET")
	if clientSecret == "" {
		clientSecret = "demo-secret"
	}

	client := axonflow.NewClient(axonflow.AxonFlowConfig{
		Endpoint:     agentURL,
		ClientID:     clientID,
		ClientSecret: clientSecret,
	})

	fmt.Println("AxonFlow HITL - require_approval Policy Example")
	fmt.Println(strings.Repeat("=", 60))

	// 1. Create a policy with require_approval action
	fmt.Println("\n1. Creating HITL oversight policy...")

	policy, err := client.CreateStaticPolicy(&axonflow.CreateStaticPolicyRequest{
		Name:        "High-Value Transaction Oversight",
		Description: "Require human approval for high-value financial decisions",
		Category:    axonflow.CategorySecurityAdmin,
		// Pattern matches amounts over 1 million (₹, $, €) - case insensitive
		Pattern:  `(?i)(amount|value|total|transaction).*[₹$€]\s*[1-9][0-9]{6,}`,
		Severity: axonflow.SeverityHigh,
		Enabled:  true,
		Action:   axonflow.ActionRequireApproval, // Triggers HITL queue
	})
	if err != nil {
		handleError(err)
	}

	fmt.Printf("   Created policy: %s\n", policy.ID)
	fmt.Printf("   Name: %s\n", policy.Name)
	fmt.Printf("   Action: %s\n", policy.Action)
	fmt.Printf("   Tier: %s\n", policy.Tier)

	// 2. Test the pattern with sample inputs
	fmt.Println("\n2. Testing pattern with sample inputs...")

	testResult, err := client.TestPattern(policy.Pattern, []string{
		"Transfer amount $5000000 to account",   // Should match (5M)
		"Transaction value ₹100000000",          // Should match (10Cr)
		"Total: €2500000",                       // Should match (2.5M)
		"Payment of $500 completed",             // Should NOT match
		"Amount: $999999",                       // Should NOT match (under 1M)
	})
	if err != nil {
		handleError(err)
	}

	fmt.Println("\n   Test results:")
	for _, match := range testResult.Matches {
		icon := "✗ PASS"
		if match.Matched {
			icon = "✓ HITL"
		}
		input := match.Input
		if len(input) > 40 {
			input = input[:40] + "..."
		}
		fmt.Printf("   %s: \"%s\"\n", icon, input)
	}

	// 3. Test enforcement via ProxyLLMCall — verify policy actually blocks
	fmt.Println("\n3. Testing HITL enforcement via ProxyLLMCall...")
	fmt.Println("   Waiting for policy propagation...")
	time.Sleep(3 * time.Second)

	userToken := os.Getenv("AXONFLOW_USER_TOKEN")

	// 3a. Send a query that MATCHES the require_approval pattern
	fmt.Println("\n   3a. Sending query that matches HITL pattern...")
	matchingResponse, matchErr := client.ProxyLLMCall(
		userToken,
		"Process transaction amount $5000000 to offshore account",
		"chat",
		map[string]interface{}{"provider": "ollama"},
	)

	if matchErr != nil {
		// In community mode, the call may succeed (auto-approve) or fail due to missing LLM key
		errStr := matchErr.Error()
		if strings.Contains(errStr, "api_key") || strings.Contains(errStr, "authentication") || strings.Contains(errStr, "API key") {
			fmt.Printf("   Note: LLM API error (expected without key): %v\n", matchErr)
			assertCheck(true, "Matching query processed (LLM key issue expected in community mode)")
		} else {
			assertCheck(false, fmt.Sprintf("Matching query failed unexpectedly: %v", matchErr))
		}
	} else if matchingResponse.Blocked {
		// Enterprise mode: policy enforcement blocks the request
		fmt.Printf("   BLOCKED: %s\n", matchingResponse.BlockReason)
		assertCheck(true, "Enterprise HITL enforcement: matching query was blocked")
		assertCheck(
			strings.Contains(matchingResponse.BlockReason, "require_approval") ||
				strings.Contains(matchingResponse.BlockReason, "approval"),
			fmt.Sprintf("Block reason mentions approval (got: %s)", matchingResponse.BlockReason),
		)
	} else {
		// Community mode: auto-approved, call succeeds
		fmt.Println("   NOT BLOCKED (community mode auto-approve)")
		assertCheck(true, "Community mode: matching query auto-approved (expected)")
	}

	// 3b. Send a safe query that should NOT trigger HITL
	fmt.Println("\n   3b. Sending safe query (should NOT trigger HITL)...")
	safeResponse, safeErr := client.ProxyLLMCall(
		userToken,
		"What is the weather today?",
		"chat",
		map[string]interface{}{"provider": "ollama"},
	)

	if safeErr != nil {
		errStr := safeErr.Error()
		if strings.Contains(errStr, "api_key") || strings.Contains(errStr, "authentication") || strings.Contains(errStr, "API key") {
			fmt.Printf("   Note: LLM API error (expected without key): %v\n", safeErr)
			assertCheck(true, "Safe query processed (LLM key issue expected)")
		} else {
			assertCheck(false, fmt.Sprintf("Safe query failed unexpectedly: %v", safeErr))
		}
	} else {
		assertCheck(!safeResponse.Blocked, "Safe query was NOT blocked by HITL policy")
	}

	// 4. Create additional HITL policies
	fmt.Println("\n4. Creating admin access oversight policy...")

	adminPolicy, err := client.CreateStaticPolicy(&axonflow.CreateStaticPolicyRequest{
		Name:        "Admin Access Detection",
		Description: "Route admin operations through human review",
		Category:    axonflow.CategorySecurityAdmin,
		Pattern:     `(admin|root|superuser|sudo|DELETE\s+FROM|DROP\s+TABLE)`,
		Severity:    axonflow.SeverityCritical,
		Enabled:     true,
		Action:      axonflow.ActionRequireApproval,
	})
	if err != nil {
		handleError(err)
	}

	fmt.Printf("   Created: %s\n", adminPolicy.Name)
	fmt.Printf("   Action: %s\n", adminPolicy.Action)

	// 5. List all policies with require_approval action
	// Note: Filter by tenant tier to get our custom policies (system policies are on earlier pages)
	fmt.Println("\n5. Listing all HITL policies...")

	tenantTier := axonflow.TierTenant
	allPolicies, err := client.ListStaticPolicies(&axonflow.ListStaticPoliciesOptions{
		Tier: tenantTier,
	})
	if err != nil {
		handleError(err)
	}

	var hitlCount int
	fmt.Println("   HITL policies:")
	for _, p := range allPolicies {
		if p.Action == axonflow.ActionRequireApproval {
			hitlCount++
			fmt.Printf("   - %s (%s)\n", p.Name, p.Severity)
		}
	}
	fmt.Printf("   Found %d HITL policies\n", hitlCount)

	// 6. Clean up test policies
	fmt.Println("\n6. Cleaning up test policies...")
	if err := client.DeleteStaticPolicy(policy.ID); err != nil {
		fmt.Printf("   Warning: Failed to delete policy: %v\n", err)
	} else {
		assertCheck(true, "Deleted high-value oversight policy")
	}
	if err := client.DeleteStaticPolicy(adminPolicy.ID); err != nil {
		fmt.Printf("   Warning: Failed to delete admin policy: %v\n", err)
	} else {
		assertCheck(true, "Deleted admin access policy")
	}

	// Final assertion summary
	fmt.Println("\n" + strings.Repeat("=", 60))
	fmt.Println("Assertion Summary")
	fmt.Println(strings.Repeat("=", 60))
	if len(failures) == 0 {
		fmt.Println("All assertions passed!")
		fmt.Println()
		fmt.Println("HITL Policy operations validated:")
		fmt.Println("  - CreateStaticPolicy() with require_approval action")
		fmt.Println("  - TestPattern() for HITL trigger validation")
		fmt.Println("  - ProxyLLMCall() enforcement (blocked or auto-approved)")
		fmt.Println("  - ListStaticPolicies() filtering by action")
		fmt.Println("  - DeleteStaticPolicy()")
		fmt.Println()
		fmt.Println("Note: In Community Edition, require_approval auto-approves.")
		fmt.Println("Upgrade to Enterprise for full HITL queue functionality.")
	} else {
		fmt.Printf("%d assertion(s) failed:\n", len(failures))
		for _, f := range failures {
			fmt.Printf("  - %s\n", f)
		}
		os.Exit(1)
	}
}

func handleError(err error) {
	fmt.Printf("\nError: %v\n", err)

	if strings.Contains(err.Error(), "connection refused") {
		fmt.Println("\nHint: Make sure AxonFlow is running:")
		fmt.Println("  docker compose up -d")
	}

	os.Exit(1)
}

File: examples/hitl/typescript/require-approval-policy.ts
/**
 * AxonFlow HITL - Create Policy with require_approval Action
 *
 * This example demonstrates how to create a policy that triggers
 * Human-in-the-Loop (HITL) approval using the `require_approval` action.
 *
 * The `require_approval` action:
 * - Enterprise: Pauses execution and creates an approval request in the HITL queue
 * - Community: Auto-approves immediately (upgrade path to Enterprise)
 *
 * Use cases:
 * - High-value transaction oversight (EU AI Act Article 14, SEBI AI/ML)
 * - Admin access detection
 * - Sensitive data access control
 *
 * VALIDATION: This example exits with code 1 if any assertion fails.
 */

import { AxonFlow } from '@axonflow/sdk';

const failures: string[] = [];

function assertCheck(condition: boolean, message: string): void {
  if (condition) {
    console.log(`   ✓ PASS: ${message}`);
  } else {
    console.log(`   ❌ FAIL: ${message}`);
    failures.push(message);
  }
}

async function main() {
  // Initialize the client with client ID for policy operations
  // clientId is used as X-Tenant-ID header for multi-tenant APIs
  const client = new AxonFlow({
    endpoint: process.env.AXONFLOW_ENDPOINT || 'http://localhost:8080',
    clientId: process.env.AXONFLOW_CLIENT_ID || 'demo-tenant',
  });

  console.log('AxonFlow HITL - require_approval Policy Example');
  console.log('='.repeat(60));

  try {
    // 1. Create a policy with require_approval action
    console.log('\n1. Creating HITL oversight policy...');

    const policy = await client.createStaticPolicy({
      name: 'High-Value Transaction Oversight',
      description: 'Require human approval for high-value financial decisions',
      category: 'security-admin',
      // Pattern matches amounts over 1 million (₹, $, €) - case insensitive
      pattern: '(?i)(amount|value|total|transaction).*[₹$€]\\s*[1-9][0-9]{6,}',
      severity: 'high',
      enabled: true,
      action: 'require_approval',  // Triggers HITL queue
    });

    console.log(`   Created policy: ${policy.id}`);
    console.log(`   Name: ${policy.name}`);
    console.log(`   Action: ${policy.action}`);
    console.log(`   Tier: ${policy.tier}`);

    assertCheck(typeof policy.id === 'string' && policy.id.length > 0, 'Policy has valid ID');
    assertCheck(policy.action === 'require_approval', 'Policy action is require_approval');
    assertCheck(policy.name === 'High-Value Transaction Oversight', 'Policy name matches input');
    assertCheck(policy.enabled === true, 'Policy is enabled');

    // 2. Test the pattern with sample inputs
    console.log('\n2. Testing pattern with sample inputs...');

    const testResult = await client.testPattern(
      policy.pattern,
      [
        'Transfer amount $5000000 to account',      // Should match (5M)
        'Transaction value ₹100000000',             // Should match (10Cr)
        'Total: €2500000',                          // Should match (2.5M)
        'Payment of $500 completed',                 // Should NOT match (under threshold)
        'Amount: $999999',                           // Should NOT match (under 1M)
      ]
    );

    console.log('\n   Test results:');
    for (const match of testResult.matches) {
      const icon = match.matched ? '✓ HITL' : '✗ PASS';
      const input = match.input.length > 40 ? match.input.substring(0, 40) + '...' : match.input;
      console.log(`   ${icon}: "${input}"`);
    }

    // Verify pattern matching behavior
    assertCheck(testResult.valid === true, 'Pattern is valid regex');
    assertCheck(testResult.matches.length === 5, 'All 5 test inputs were evaluated');
    // First 3 should match (over 1M), last 2 should not match (under 1M)
    assertCheck(testResult.matches[0].matched === true, 'High-value $5M matches pattern');
    assertCheck(testResult.matches[1].matched === true, 'High-value 10Cr INR matches pattern');
    assertCheck(testResult.matches[2].matched === true, 'High-value 2.5M EUR matches pattern');
    assertCheck(testResult.matches[3].matched === false, 'Low-value $500 does not match pattern');
    assertCheck(testResult.matches[4].matched === false, 'Under-threshold $999999 does not match pattern');

    // 3. Test enforcement via proxyLLMCall — verify policy actually blocks
    console.log('\n3. Testing HITL enforcement via proxyLLMCall...');
    console.log('   Waiting for policy propagation...');
    await new Promise(resolve => setTimeout(resolve, 3000));

    const userToken = process.env.AXONFLOW_USER_TOKEN || '';

    // 3a. Send query that MATCHES the require_approval pattern
    console.log('\n   3a. Sending query that matches HITL pattern...');
    try {
      const matchingResponse = await client.proxyLLMCall({
        userToken,
        query: 'Process transaction amount $5000000 to offshore account',
        requestType: 'chat',
        context: { provider: 'anthropic' },
      });

      if (matchingResponse.blocked) {
        // Enterprise mode: policy enforcement blocks the request
        assertCheck(true, 'Enterprise HITL enforcement: matching query was blocked');
        const blockReason = matchingResponse.blockReason || '';
        assertCheck(
          blockReason.includes('require_approval') || blockReason.includes('approval'),
          `Block reason mentions approval (got: ${blockReason})`
        );
      } else {
        // Community mode: auto-approved
        assertCheck(true, 'Community mode: matching query auto-approved (expected)');
      }
    } catch (error) {
      const errMsg = error instanceof Error ? error.message.toLowerCase() : String(error).toLowerCase();
      if (errMsg.includes('api_key') || errMsg.includes('authentication')) {
        console.log(`   Note: LLM API error (expected without key): ${error instanceof Error ? error.message : error}`);
        assertCheck(true, 'Matching query processed (LLM key issue expected)');
      } else {
        assertCheck(false, `Matching query failed unexpectedly: ${error instanceof Error ? error.message : error}`);
      }
    }

    // 3b. Send safe query that should NOT trigger HITL
    console.log('\n   3b. Sending safe query (should NOT trigger HITL)...');
    try {
      const safeResponse = await client.proxyLLMCall({
        userToken,
        query: 'What is the weather today?',
        requestType: 'chat',
        context: { provider: 'anthropic' },
      });
      assertCheck(!safeResponse.blocked, 'Safe query was NOT blocked by HITL policy');
    } catch (error) {
      const errMsg = error instanceof Error ? error.message.toLowerCase() : String(error).toLowerCase();
      if (errMsg.includes('api_key') || errMsg.includes('authentication')) {
        console.log(`   Note: LLM API error (expected without key): ${error instanceof Error ? error.message : error}`);
        assertCheck(true, 'Safe query processed (LLM key issue expected)');
      } else {
        assertCheck(false, `Safe query failed unexpectedly: ${error instanceof Error ? error.message : error}`);
      }
    }

    // 4. Create additional HITL policies for different use cases
    console.log('\n4. Creating admin access oversight policy...');

    const adminPolicy = await client.createStaticPolicy({
      name: 'Admin Access Detection',
      description: 'Route admin operations through human review',
      category: 'security-admin',
      pattern: '(admin|root|superuser|sudo|DELETE\\s+FROM|DROP\\s+TABLE)',
      severity: 'critical',
      enabled: true,
      action: 'require_approval',
    });

    console.log(`   Created: ${adminPolicy.name}`);
    console.log(`   Action: ${adminPolicy.action}`);

    assertCheck(adminPolicy.action === 'require_approval', 'Admin policy has require_approval action');
    assertCheck(adminPolicy.severity === 'critical', 'Admin policy severity is critical');

    // 5. List all policies with require_approval action
    console.log('\n5. Listing all HITL policies...');

    // Include tenant-tier policies with limit parameter
    const allPolicies = await client.listStaticPolicies({ limit: 100 });
    const hitlPolicies = allPolicies.filter(p => p.action === 'require_approval');

    console.log(`   Found ${hitlPolicies.length} HITL policies:`);
    for (const p of hitlPolicies) {
      console.log(`   - ${p.name} (${p.severity})`);
    }

    assertCheck(hitlPolicies.length >= 2, 'At least 2 HITL policies exist (created in this test)');
    assertCheck(
      hitlPolicies.some(p => p.name === 'High-Value Transaction Oversight'),
      'High-Value Transaction Oversight policy found in HITL list'
    );
    assertCheck(
      hitlPolicies.some(p => p.name === 'Admin Access Detection'),
      'Admin Access Detection policy found in HITL list'
    );

    // 6. Clean up test policies
    console.log('\n6. Cleaning up test policies...');
    await client.deleteStaticPolicy(policy.id);
    await client.deleteStaticPolicy(adminPolicy.id);
    console.log('   Deleted test policies');
    assertCheck(true, 'Test policies deleted successfully');

    // Verify policies are deleted
    const remainingPolicies = await client.listStaticPolicies();
    const deletedPolicyExists = remainingPolicies.some(p => p.id === policy.id || p.id === adminPolicy.id);
    assertCheck(!deletedPolicyExists, 'Deleted policies no longer exist in list');

    console.log('\n' + '='.repeat(60));
    console.log('Example completed successfully!');
    console.log('\nNote: In Community Edition, require_approval auto-approves.');
    console.log('Upgrade to Enterprise for full HITL queue functionality.');

  } catch (error) {
    if (error instanceof Error) {
      console.error('\nError:', error.message);

      if (error.message.includes('ECONNREFUSED')) {
        console.error('\nHint: Make sure AxonFlow is running:');
        console.error('  docker compose up -d');
      }
    }
    failures.push(`Unexpected error: ${error instanceof Error ? error.message : error}`);
  }

  // Final assertion summary
  console.log('\n' + '='.repeat(60));
  console.log('Assertion Summary');
  console.log('='.repeat(60));
  if (failures.length === 0) {
    console.log('All assertions passed!');
  } else {
    console.log(`${failures.length} assertion(s) failed:`);
    failures.forEach((f) => console.log(`  - ${f}`));
  }

  process.exit(failures.length > 0 ? 1 : 0);
}

main();
